mlflow/store/ftp_artifact_repo.py

```diff
@@ -0,0 +1,124 @@
+import ftplib
+import os
+from ftplib import FTP
+
+from six.moves import urllib
+
+from mlflow.entities.file_info import FileInfo
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.utils.file_utils import TempDir, build_path, get_relative_path
+
+
+class FTPArtifactRepository(ArtifactRepository):
+    """Stores artifacts as files in a remote directory, via sftp."""
+
+    def __init__(self, artifact_uri, client=None):
+        self.uri = artifact_uri
+        parsed = urllib.parse.urlparse(artifact_uri)
+        self.config = {
+            'host': parsed.hostname,
+            'port': 21 if parsed.port is None else parsed.port,
+            'username': parsed.username,
+            'password': parsed.password
+        }
+        self.path = parsed.path
+
+        if self.config['host'] is None:
+            self.config['host'] = 'localhost'
+
+        if client:
+            self.ftp = client
+        else:
+            self.ftp = FTP()
+            self.ftp.connect(self.config['host'], self.config['port'])
+            self.ftp.login(self.config['username'], self.config['password'])
+
+        super(FTPArtifactRepository, self).__init__(artifact_uri)
+
+    def _is_dir(self, full_file_path):
+        try:
+            self.ftp.cwd(full_file_path)
+            return True
+        except ftplib.error_perm:
+            return False
+
+    def _mkdir(self, artifact_dir):
+        try:
+            self.ftp.mkd(artifact_dir)
+        except ftplib.error_perm:
+            head, _ = os.path.split(artifact_dir)
+            self._mkdir(head)
+            self._mkdir(artifact_dir)
+
+    def download_files(self, path, destination):
+        self.ftp.cwd(path)
+        if not os.path.isdir(destination):
+            os.makedirs(destination)
+
+        filelist = self.ftp.nlst()
+
+        for ftp_file in filelist:
+            if self._is_dir(build_path(path, ftp_file)):
+                self.download_files(build_path(path, ftp_file), build_path(destination, ftp_file))
+            else:
+                with open(os.path.join(destination, ftp_file), "wb") as f:
+                    self.ftp.retrbinary("RETR "+ftp_file, f)
+        return
+
+    def log_artifact(self, local_file, artifact_path=None):
+        artifact_dir = os.path.join(self.path, artifact_path) \
+            if artifact_path else self.path
+        self._mkdir(artifact_dir)
+        with open(local_file, 'rb') as f:
+            self.ftp.cwd(artifact_dir)
+            self.ftp.storbinary('STOR ' + os.path.split(local_file)[1], f)
```

Comment:
Nit: for clarity, let's use `os.path.basename` instead of `os.path.split(local_file)[1]`

Reply:
fixed